helix-view/src/document.rs

```diff
@@ -982,6 +989,21 @@ impl Document {
         server.is_initialized().then(|| server)
     }
 
+    pub fn differ(&self) -> Option<&Differ> {
+        self.differ.as_ref()
+    }
+
+    /// Intialize/updates the differ for this document with a new base.
+    pub fn set_diff_base(&mut self, diff_base: Vec<u8>) {
+        if let Ok((diff_base, _)) = from_reader(&mut diff_base.as_slice(), Some(self.encoding)) {
+            if let Some(differ) = &self.differ {
+                differ.update_diff_base(diff_base);
+                return;
+            }
+            self.differ = Some(Differ::new(diff_base, self.text.clone()))
```

Comment:
Is it a good idea to have the whole doc cloned just for diff? Means we need to keep the whole text twice? Once for editor and one just for diff?

Reply:
`Rope`s internally use `Arc`s, so it's cheap to clone one.

Reply:
As @kirawi said a `Rope` is very cheap to clone and -as far as I am aware- designed exactly for this kind of situation (sending a snapshot to another thread and processing it asynchronously)

Regarding using something else as a `diff_base` I did consider sending a `String` instead of `Rope` but I choose the rope for two reasons:

* it allowed me to reuse the existing decoding implementation without too many additional changes (so keeping the PR minimal)
* more importantly: I hope to reuse the `Differ` to implement a git difftool like side by side comparison (not just between a git revision and an open file but also between two open files for example) in the (probably somewhat distant) future. In that case`diff_base` is actually also an open document that would be represented as a rope anyway. It will take quite a while to implement something like that (and even longer to get it merged) so for now we could use a string for slightly better performance but I would think that the overhead is small enough that we can justify keeping it as a `Rope` in the name of future proofing the implementation.

What do you think?

Reply:
Yeah that's fine.